loader/lightning.go

```diff
@@ -0,0 +1,306 @@
+	lightningCfg.TiDB.PdAddr = cfg.TiDB.PdAddr
+	lightningCfg.TikvImporter.Backend = cfg.TiDB.Backend
+	lightningCfg.PostRestore.Checksum = lcfg.OpLevelOff
+	if cfg.TiDB.Backend == lcfg.BackendLocal {
+		lightningCfg.TikvImporter.SortedKVDir = cfg.Dir
+	}
+	lightningCfg.Mydumper.SourceDir = cfg.Dir
+	return lightningCfg
+}
+
+// Type implements Unit.Type.
+func (l *LightningLoader) Type() pb.UnitType {
+	return pb.UnitType_Load
+}
+
+// Init initializes loader for a load task, but not start Process.
+// if fail, it should not call l.Close.
+func (l *LightningLoader) Init(ctx context.Context) (err error) {
+	tctx := tcontext.NewContext(ctx, l.logger)
+	checkpoint, err := newRemoteCheckPoint(tctx, l.cfg, l.checkpointID())
+	failpoint.Inject("ignoreLoadCheckpointErr", func(_ failpoint.Value) {
+		l.logger.Info("", zap.String("failpoint", "ignoreLoadCheckpointErr"))
+		err = nil
+	})
+	l.checkPoint = checkpoint
+	l.toDB, l.toDBConns, err = createConns(tctx, l.cfg, 1)
+	return err
+}
+
+func (l *LightningLoader) restore(ctx context.Context) error {
+	tctx := tcontext.NewContext(ctx, l.logger)
+	if err := putLoadTask(l.cli, l.cfg, l.workerName); err != nil {
+		return err
+	}
+	if err := l.checkPoint.Init(tctx, lightningCheckpointFile, 1); err != nil {
+		return err
+	}
+	if err := l.checkPoint.Load(tctx); err != nil {
+		return err
+	}
+	db2Tables := make(map[string]Tables2DataFiles)
+	tables := make(Tables2DataFiles)
+	files := make(DataFiles, 0, 1)
+	files = append(files, lightningCheckpointFile)
+	tables[lightningCheckpointTable] = files
+	db2Tables[lightningCheckpointDB] = tables
+	var err error
+	if err = l.checkPoint.CalcProgress(db2Tables); err != nil {
+		return err
+	}
+	if !l.checkPoint.IsTableFinished(lightningCheckpointDB, lightningCheckpointTable) {
+		cfg := lcfg.NewConfig()
+		if err = cfg.LoadFromGlobal(l.lightningConfig); err != nil {
+			return err
+		}
+		cfg.Routes = l.cfg.RouteRules
+		cfg.Checkpoint.Schema = "tidb_lightning_checkpoint_" + l.cfg.Name
+		cfg.Checkpoint.Driver = lcfg.CheckpointDriverMySQL
+		param := common.MySQLConnectParam{
+			Host:             cfg.TiDB.Host,
+			Port:             cfg.TiDB.Port,
+			User:             cfg.TiDB.User,
+			Password:         cfg.TiDB.Psw,
+			SQLMode:          mysql.DefaultSQLMode,
+			MaxAllowedPacket: 64 * units.MiB,
+			TLS:              cfg.TiDB.TLS,
+		}
+		cfg.Checkpoint.DSN = param.ToDSN()
+		if err = cfg.Adjust(ctx); err != nil {
+			return err
+		}
+		l.Lock()
+		taskCtx, cancel := context.WithCancel(ctx)
+		l.cancel = cancel
+		l.Unlock()
+		err = l.core.RunOnce(taskCtx, cfg, nil)
+		if err == nil {
+			l.finish.Store(true)
+			offsetSQL := l.checkPoint.GenSQL(lightningCheckpointFile, 1)
+			err = l.toDBConns[0].executeSQL(tctx, []string{offsetSQL})
+			_ = l.checkPoint.UpdateOffset(lightningCheckpointFile, 1)
+		}
+	} else {
+		l.finish.Store(true)
+	}
+	if l.cfg.Mode == config.ModeFull {
+		_ = delLoadTask(l.cli, l.cfg, l.workerName)
```

Comment:
I think this is the only chance to `delLoadTask`, so we'd better let the task fail and retry to delete the marker in etcd. Otherwise the source is always pinned to this worker.